js/libs/keycloak-admin-client/src/utils/auth.ts

```diff
@@ -77,9 +80,10 @@ export const getToken = async (settings: Settings): Promise<TokenResponse> => {
   const headers = new Headers(options.headers);
 
   if (credentials.clientSecret) {
+    // See: https://datatracker.ietf.org/doc/html/rfc2617#section-2
     headers.set(
       "Authorization",
-      `Basic ${btoa(`${credentials.clientId}:${credentials.clientSecret}`)}`
+      `Basic ${toBase64(`${credentials.clientId}:${credentials.clientSecret}`)}`
```

Comment:
@sschu from what I could find in [the spec](https://datatracker.ietf.org/doc/html/rfc2617#section-2) this whole string should be encoded.

Reply:
Exactly, but clientId and clientSecret also have to be URL-encoded according to https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1

Reply:
I am having a hard time understanding this spec, so let's break this down:

> Clients in possession of a client password MAY use the HTTP Basic authentication scheme as defined in [[RFC2617](https://datatracker.ietf.org/doc/html/rfc2617)] to authenticate with the authorization server.

So this seems to be correct so far. 

> The client identifier is encoded using the "application/x-www-form-urlencoded" encoding algorithm per [Appendix B](https://datatracker.ietf.org/doc/html/rfc6749#appendix-B), and the encoded value is used as the username; The client password is encoded using the same algorithm and used as the password.

So this pertains to `credentials.clientId` and `credentials.clientSecret` in our code. And here is where we get buried in a bunch of specs that stipulate it's encoding:

> To address this shortcoming when generating payloads using this media type, names and values MUST be encoded using the UTF-8 character encoding scheme [[RFC3629](https://datatracker.ietf.org/doc/html/rfc3629)] first; the resulting octet sequence then needs to be further encoded using the escaping rules defined in [[W3C.REC-html401-19991224](https://datatracker.ietf.org/doc/html/rfc6749#ref-W3C.REC-html401-19991224)].

So this is where it breaks down for me, looking at the [Java code](https://github.com/keycloak/keycloak/blob/fbfdb547459dfa4f82256a60aef03f6b32dba1dd/core/src/main/java/org/keycloak/util/BasicAuthHelper.java#L73-L76), I would expect something as follows:

```ts
// See: https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1
const username = encodeURIComponent(credentials.clientId);
const password = encodeURIComponent(credentials.clientSecret);

// See: https://datatracker.ietf.org/doc/html/rfc2617#section-2
headers.set(
  "authorization",
  `Basic ${toBase64(`${username}:${password}`)}`
);
```

This does seem to cover the UTF-8 conversion according to MDN:

> The `encodeURIComponent()` function encodes a [URI](https://developer.mozilla.org/en-US/docs/Glossary/URI) by replacing each instance of certain characters by one, two, three, or four escape sequences representing the [UTF-8](https://developer.mozilla.org/en-US/docs/Glossary/UTF-8) encoding of the character (will only be four escape sequences for characters composed of two surrogate characters).

However, it looks like there are s[ome implementation differences](https://stackoverflow.com/questions/607176/java-equivalent-to-javascripts-encodeuricomponent-that-produces-identical-outpu#answer-607403) with the Java code. Which leaves me wondering, which one of these conforms to the aforementioned specs?

WDYT @sschu?

Reply:
I did not expect this to be so complicated, but here we are...

In my understanding, the Java code does exactly what `x-www-form-urlencoded`requires. 

Looking at the differences, I assume ` ` has to be encoded as `+` additionally in Javascript, since "For [application/x-www-form-urlencoded](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#application/x-www-form-urlencoded-encoding-algorithm), spaces are to be replaced by +, so one may wish to follow a encodeURIComponent() replacement with an additional replacement of %20 with +." (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent).

The other difference between Java and JS seems to be that JS would not encode `[~'()!]`. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#encoding_for_rfc3986 shows a piece of code how to do that. So maybe this could be used? However, although it returns the UTF-16 byte representations of characters (`.charCodeAt(0)`) they seem to be the same as UTF-8.

Reply:
> I did not expect this to be so complicated, but here we are...

Hahah, yeah. But let's solve this properly. Sorry for the late reply, I had to take care of some other tasks.

> Looking at the differences, I assume ` ` has to be encoded as `+` additionally in JavaScript

Yes, from what I have read this indeed seems to be the case. I've added a function to take care of this additional encoding. As well as a new method that handles encoding in Base64 a bit better.

> The other difference between Java and JS seems to be that JS would not encode `[~'()!]`. So maybe [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#encoding_for_rfc3986) could be used?

I am not sure that we need to conform to RFC3986 here, I can't seem to find a reference of it under [Appendix B](https://datatracker.ietf.org/doc/html/rfc6749#appendix-B). But I guess it can do no harm? I'll add this.

Reply:
@sschu I am fairly confident this now matches the spec, do you agree?

Reply:
Awesome, then I will consider this resolved.